youtube/services.py

```diff
@@ -0,0 +1,83 @@
+import requests
+from .models import Video
+from core.utils import db_table_exists
+from django.utils import timezone
+from core.models import User
+
+
+class YoutubeService:
+
+    @staticmethod
+    def get_user_channel_id(username, api_key):
+        r = requests.get('https://www.googleapis.com/youtube/v3/channels?part=id&forUsername=' + username
+                         + '&key=' + api_key)
+        r = r.json()
+        if not r['items']:
+            return None
+        return r['items'][0]['id']
+
+    @staticmethod
+    def get_my_videos(username, api_key, max_results=15):
+        if not(db_table_exists('youtube_video')):
+            return None
+        my_videos = []
+        if not User.is_authenticated:
+            return None
+        r = requests.get('https://www.googleapis.com/youtube/v3/channels?part=contentDetails&forUsername=' +
+                         username + '&key=' + api_key)
+        r = r.json()
+        if not r['items']:
+            return None
+        my_uploaded_playlist = r['items'][0]['contentDetails']['relatedPlaylists']['uploads']
+        r = requests.get('https://www.googleapis.com/youtube/v3/playlistItems?part=snippet&maxResults=' +
+                         str(max_results) + '&playlistId=' + my_uploaded_playlist + '&key=' + api_key)
+        r = r.json()
+        for item in r['items']:
+            youtube_id = item['snippet']['resourceId']['videoId']
+            published_at = item['snippet']['publishedAt']
+            channel_id = item['snippet']['channelId']
+            title = item['snippet']['title']
+            description = item['snippet']['description']
+            thumbnail = item['snippet']['thumbnails']['medium']['url']
+            channel_title = item['snippet']['channelTitle']
+            if Video.objects.filter(youtube_id=youtube_id, title=title).exists():
```

Comment:
dodałeś sprawdzanie filmików w `for`, to nie jest dobre rozwiązanie, bo będziesz robił dożu zapytań
ja bym to zrobił w taki sposób
przed `for` stworzyć listę `videos_to_add = []`, w `for` dodajesz obiekty do tej listy, czyli
```
videos_to_add.append(Video(youtube_id=youtube_id, published_at=published_at, channel_id=channel_id,
                                         title=title, description=description, thumbnail=thumbnail,
                                        channel_title=channel_title))
```
dalej jak już `for` sie skończy, to szukasz filmiki w bazie używając `IN`, i usuwasz z listy istniejące objekty, te obiekty co zostały w liscie dodajesz do bazy 1 zapytaniem, używając `bulk_create` (https://docs.djangoproject.com/en/2.0/ref/models/querysets/#bulk-create)

Reply:
i jeszcze pytania co z filmikami co już są `expired`? jeżeli nadal są popularne, to musisz zmienić `expiration_date` dla nich